pkg/provider/kubernetes/gateway.go

```diff
@@ -0,0 +1,167 @@
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	gwapiv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/envoyproxy/gateway/pkg/envoygateway/config"
+)
+
+type gatewayReconciler struct {
+	client client.Client
+	// classController is the configured gatewayclass controller name.
+	classController gwapiv1b1.GatewayController
+	log             logr.Logger
+}
+
+// newGatewayController creates a gateway controller. The controller will watch for
+// Gateway objects across all namespaces and reconcile those that match the configured
+// gatewayclass controller name.
+func newGatewayController(mgr manager.Manager, cfg *config.Server) error {
+	r := &gatewayReconciler{
+		client:          mgr.GetClient(),
+		classController: gwapiv1b1.GatewayController(cfg.EnvoyGateway.Gateway.ControllerName),
+		log:             cfg.Logger,
+	}
+
+	c, err := controller.New("gateway", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+	r.log.Info("created gateway controller")
+
+	// Only enqueue Gateway objects that match this Envoy Gateway's controller name.
+	if err := c.Watch(
+		&source.Kind{Type: &gwapiv1b1.Gateway{}},
+		&handler.EnqueueRequestForObject{},
+		predicate.NewPredicateFuncs(r.hasMatchingController),
+	); err != nil {
+		return err
+	}
+	r.log.Info("watching gateway objects")
+
+	return nil
+}
+
+// hasMatchingController returns true if the provided object is a Gateway
+// using a GatewayClass matching the configured gatewayclass controller name.
+func (r *gatewayReconciler) hasMatchingController(obj client.Object) bool {
+	gw, ok := obj.(*gwapiv1b1.Gateway)
+	if !ok {
+		r.log.Info("unexpected object type, bypassing reconciliation", "object", obj)
+		return false
+	}
+
+	gc := &gwapiv1b1.GatewayClass{}
+	key := types.NamespacedName{Name: string(gw.Spec.GatewayClassName)}
+	if err := r.client.Get(context.Background(), key, gc); err != nil {
+		r.log.Error(err, "failed to get gatewayclass", "name", gw.Spec.GatewayClassName)
+		return false
+	}
+
+	if gc.Spec.ControllerName != r.classController {
+		r.log.Info("gatewayclass name for gateway doesn't match configured name",
+			"namespace", gw.Namespace, "name", gw.Name)
+		return false
+	}
+
+	return true
+}
+
+// Reconcile finds all the Gateways for the GatewayClass with an "Accepted: true" condition.
+// It passes the oldest such Gateway to the IR for processing, and sets a "Scheduled: false"
+// condition on all other Gateways for the accepted GatewayClass.
```

Comment:
Godoc should be updated, we're passing _all_ Gateways for the chosen GatewayClass along for processing, not just the oldest one.